app/serializers/api/v1/package_serializer.rb

```diff
@@ -39,6 +39,14 @@ def sent_on__sql
       "stockit_sent_on"
     end
 
+    def available_quantity
+      object.total_available_quantity
+    end
+
+    def available_quantity__sql
+      "(packages.received_quantity - ((SELECT COALESCE(sum(quantity), 0) FROM orders_packages WHERE orders_packages.state = 'designated' AND orders_packages.package_id = packages.id) - (SELECT COALESCE(sum(dispatched_quantity), 0) FROM orders_packages WHERE orders_packages.state = 'designated' AND orders_packages.package_id = packages.id)))"
```

Comment:
Can I confirm that the equation should be `available_quantity = received_quantity - designated_quantity - dispatched_quantity`? If so, the SQL isn't quite right here... we need:

- quantity from designated orders_packages
- dispatched_quantity from dispatched orders_packages

```suggestion
      "(packages.received_quantity - ((SELECT COALESCE(sum(quantity), 0) FROM orders_packages WHERE orders_packages.state = 'designated' AND orders_packages.package_id = packages.id) - (SELECT COALESCE(sum(dispatched_quantity), 0) FROM orders_packages WHERE orders_packages.state = 'dispatched' AND orders_packages.package_id = packages.id)))"
```

Reply:
This SQL is formed by referring following code 

```ruby
        query = OrdersPackage.designated.where(package: package)
        query = query.where(order: to_order) if to_order.present?
        query.reduce(0) { |sum, op| sum + op.quantity - op.dispatched_quantity }

```
https://github.com/crossroads/api.goodcity/blob/master/app/models/concerns/inventory_computer.rb#L44-L48

The dispatched quantity considered here is from the partial dispatch, and hence we are looking into `designated` state only. 

cc: @patrixr